### PR TITLE
Fix '0 IF 0x01 0x81 ENDIF 1 '  ✅

### LINKED ISSUE #131: [bug] [bitcoin-core test] '0 IF 0x01 0x81 ENDIF 1 '
### Issue

The following bitcoin-core test is failing to give the correct result :
```
0 IF 0x01 0x81 ENDIF 1
```

We expect this test to pass and execute successfully.
More info from the test: `1NEGATE equiv`

### Replicate

To replicate the issue you can use the latest `main` branch with the following command:
```
scarb cairo-run '[[],18006118492099244450712950776978963358421403790876721,22,[],0,0]'
```

### PR DESCRIPTION
**Pull request by jsandinoDev**
- ✅ issue #131 
- ✅ follows contribution [guide](https://github.com/keep-starknet-strange/shinigami/blob/main/CONTRIBUTING.md)
- ✅ code change includes tests

1.  Add the **_is_pushing_opcode_** function in the _**flow.cairo**_ file.This function is to verify if the OP_CODE has the function of pushing the next OP_CODE in the stack. (0x01-0x4b)
2. Add if validation in the **_execute_** function of the **_engine_**.cairo file.
3. If the OP_CODE is a pushing op_code. The verification of is_opcode_disabled is omitted because is not required. 
4. If the OP_CODE is not a pushing op_code. Continue with the verification of the is_opcode_disabled.

(After commits of Agu 22, the results continue to be the same as in the image)

Build
Screenshot 2024-08-21 at 8 14 51 PM

Test
Screenshot 2024-08-21 at 8 15 40 PM

Closes #131 
Screenshot 2024-08-21 at 8 16 29 PM

### CODE REVIEW

File: src/engine.cairo [hunk 1]
```diff
@@ -137,13 +137,18 @@ pub impl EngineImpl of EngineTrait {
                 }
 
                 if !self.cond_stack.branch_executing() && !flow::is_branching_opcode(opcode) {
-                    let res = Opcode::is_opcode_disabled(opcode, ref self);
-                    if res.is_err() {
-                        err = res.unwrap_err();
-                        break;
+                    if flow::is_pushing_opcode(opcode) {
+                        self.opcode_idx += 2;
```
Comment (b-j-roberts) [inline src/engine.cairo:141]: The push opcode wont always contain 2 bytes of data, it depends on the opcode.
I recommend looking at OP_DATA_X and OP_PUSHDATAX opcodes for more info

File: src/opcodes/flow.cairo [hunk 2]
```diff
@@ -14,6 +14,13 @@ pub fn is_branching_opcode(opcode: u8) -> bool {
     return false;
 }
 
+pub fn is_pushing_opcode(opcode: u8) -> bool {
+    if opcode == Opcode::OP_DATA_1 || opcode == Opcode::OP_DATA_75 {
```
Comment (b-j-roberts) [inline src/opcodes/flow.cairo:18]: These aren't all the pushing opcodes, you might want to do a range check too.

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

File: src/opcodes/flow.cairo (hunk 2 above)
Comment (jsandinoDev) [reply src/opcodes/flow.cairo:18]: In the new commit I created the funcion is_data_code to separate the pushing OP_CODES based on their functionality.

- The is_data_opcode for the OP_CODE that just pushes the next bytecode in the stack

- The is_pushing_opcode to the OP_code that the next OP_CODE contains the quantity of bytes

File: src/engine.cairo (hunk 1 above)
Comment (jsandinoDev) [reply src/engine.cairo:151]: To verify all the posibilities of how many bytes are gone be push I refactor the validations
I use this case as example (If it is wrong please let me know please)

0x01  0x4c 0x04 0x11223344

1 - The IF of the line 139 validates that the op_code is OP_DATA_1 or OP_DATA_75, this OP_CODES just push the next bytes into the array.
2 - The IF of the line 141 validates if  next_opcode is a pushing_opcode.
     - If the result is true: the engine will verify if the next_opcode(2) is disable and then update the index and continue.
     - If the result is false: the index will update an ommit the verification of the op that is going to be add to the stack.

**Comment by b-j-roberts** [comment]
I'd recommend reading this to get a better understanding of how OP_DATA_X and OP_PUSHDATAX work https://wiki.bitcoinsv.io/index.php/Pushdata_Opcodes
The current implementation isn't handling how the opcodes are used fully correctly

**Pull request closed without merging**